### PR TITLE
feat: Added Athena Connector

### PR DESCRIPTION
**Pull request by TivonB-AI2**
## Description

Added AWS Athena Connector for onboarding

## Related Issue

https://linear.app/ai-squared/issue/AIS-25/day-2-4-gain-practical-experience-with-multiwoven-integrations 

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] New Connector (Destination or Source Connector)
- [ ] Breaking change (fix or feature that would impact existing functionality)
- [ ] Styling change
- [ ] Documentation update
- [ ] Refactoring
- [ ] Chore

## How Has This Been Tested?

Using Ruby on Rails to check the connector based on multiwoven protocol

## Checklist:

- [x] Ensure a branch name is prefixed with `feature`, `bugfix`, `hotfix`, `release`, `style` or `chore` followed by `/` and branch name e.g `feature/add-salesforce-connector`
- [x] Added unit tests for the changes made (if required)
- [ ] Have you made sure the commit messages meets the guidelines?
- [ ] Added relevant screenshots for the changes
- [ ] Have you tested the changes on local/staging?
- [ ] Added the new connector in rollout.rb
- [ ] Have you updated the version number of the gem?
- [ ] Have you ensured that your changes for new connector are documented in the [docs repo](https://github.com/Multiwoven/docs) or relevant documentation files?
- [ ] Have you updated the run time dependency in multiwoven-integrations.gemspec if new gems are added
- [ ] Have you made sure the code you have written follows the best practises to the best of your knowledge?

### CODE REVIEW

File: integrations/spec/multiwoven/integrations/source/athena/client_spec.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
+# frozen_string_literal: true
```
Comment (afthabvp) [inline integrations/spec/multiwoven/integrations/source/athena/client_spec.rb:1]: the client_spec is wrong.  
please read how to write test([rspec](https://github.com/rspec/rspec-core)) 
reference: 
https://github.com/Multiwoven/multiwoven/blob/main/integrations/spec/multiwoven/integrations/source/postgresql/client_spec.rb

Comment (TivonB-AI2) [reply integrations/spec/multiwoven/integrations/source/athena/client_spec.rb:1]: client_spec has been modified/corrected.

File: integrations/lib/multiwoven/integrations/source/athena/client.rb [hunk 2]
```diff
@@ -0,0 +1,117 @@
+# frozen_string_literal: true
+
+require "aws-sdk-athena"
+
+module Multiwoven::Integrations::Source
+  module AWSAthena
```
Comment (afthabvp) [inline integrations/lib/multiwoven/integrations/source/athena/client.rb:6]: nitpick: We follow rails convention while creating folders in integration gem i.e if the class name is AWSAthena we name the folder as aws_athena.
in rails it will throw an error via zeitwerk gem.

File: integrations/lib/multiwoven/integrations/source/athena/client.rb [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        handle_exception(
+          "AWS:ATHENA:READ:EXCEPTION",
+          "error",
+          e
+        )
+      end
+
+      private
+
+      def create_connection(connection_config)
```
Comment (afthabvp) [inline integrations/lib/multiwoven/integrations/source/athena/client.rb:68]: Add validation for the Athena client. Throw an error if the credentials are invalid.

File: integrations/lib/multiwoven/integrations/source/athena/client.rb [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        Aws.config.update({ credentials: Aws::Credentials.new(connection_config[:access_key], connection_config[:secret_access_key]), region: "us-east-2" })
```
Comment (afthabvp) [inline integrations/lib/multiwoven/integrations/source/athena/client.rb:69]: Take the input "region: "us-east-2"" from the connection spec. No need to hardcode i

File: integrations/lib/multiwoven/integrations/source/athena/client.rb [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        connection_config = connection_config.with_indifferent_access
+        query = "SELECT table_name, column_name, data_type, is_nullable FROM information_schema.columns WHERE table_schema = '#{connection_config[:schema]}' ORDER BY table_name, ordinal_position;"
+
+        db = create_connection(connection_config)
+        response = db.start_query_execution(
+          query_string: query,
+          result_configuration: { output_location: connection_config[:output_location] }
+        )
+        query_execution_id = response[:query_execution_id]
+        # Polling for query execution completion
```
Comment (afthabvp) [inline integrations/lib/multiwoven/integrations/source/athena/client.rb:27]: The discover and read query fetch is being done asynchronously. How will we get a response for each call? Can we do it synchronously?

File: integrations/lib/multiwoven/integrations/source/athena/client.rb [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        connection_config = connection_config.with_indifferent_access
+        query = sync_config.model.query
+        query = batched_query(query, sync_config.limit, sync_config.offset) unless sync_config.limit.nil? && sync_config.offset.nil?
+
+        db = create_connection(connection_config)
+        response = db.start_query_execution(
+          query_string: query,
+          result_configuration: { output_location: sync_config[:source][:connection_specification][:output_location] }
+        )
+        query_execution_id = response[:query_execution_id]
```
Comment (afthabvp) [inline integrations/lib/multiwoven/integrations/source/athena/client.rb:53]: same here. async

File: integrations/lib/multiwoven/integrations/source/athena/client.rb (hunk 2 above)
Comment (subintp) [reply integrations/lib/multiwoven/integrations/source/athena/client.rb]: Folder name should be same as class name. @TivonB-AI2 Refer other connectors

Comment (TivonB-AI2) [reply integrations/lib/multiwoven/integrations/source/athena/client.rb]: They should now be renamed to aws_athena

File: integrations/lib/multiwoven/integrations/source/aws_athena/config/spec.json [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        "examples": ["AWSATHENAEXAMPLE"],
+        "type": "string",
+        "title": "Personal Access Key",
+        "order": 0
+      },
+      "secret_access_key": {
+        "description": "The AWS Secret Access Key to use for authentication.",
+        "examples": ["AWS/ATHENA/EXAMPLEKEY"],
+        "type": "string",
+        "title": "Secret Access Key",
```
Comment (subintp) [inline integrations/lib/multiwoven/integrations/source/aws_athena/config/spec.json:22]: @TivonB-AI2  add multiwoven_secret to true to mask this in the ui for credentails. refer other connectors

**Review by subintp — CHANGES_REQUESTED** [review]

File: integrations/lib/multiwoven/integrations/source/aws_athena/config/spec.json (hunk 7 above)
Comment (TivonB-AI2) [reply integrations/lib/multiwoven/integrations/source/aws_athena/config/spec.json:23]: Done

File: integrations/Gemfile [hunk 8]
```diff
@@ -41,7 +41,7 @@ gem "async-websocket", "~> 0.8.0"
 
 gem "git"
 
-gem "hubspot-api-client"
+gem "hubspot-api-client", "~> 17.2.0"
```
Comment (karthik-sivadas) [inline integrations/Gemfile:44]: Why do we need this change and how is this related with AWS Athena connector

File: integrations/lib/multiwoven/integrations/source/aws_athena/client.rb [hunk 9]
```diff
@@ -0,0 +1,121 @@
+# frozen_string_literal: true
+
+require "aws-sdk-athena"
```
Comment (karthik-sivadas) [inline integrations/lib/multiwoven/integrations/source/aws_athena/client.rb:3]: Please add this in `integrations/lib/multiwoven/integrations.rb`

File: integrations/Gemfile (hunk 8 above)
Comment (TivonB-AI2) [reply integrations/Gemfile:44]: I needed to add aws-sdk-athena but if I gem install hubspot will become a version higher than 17.2.0 causing the hubspot client_rspec to fail. In short this is used to prevent the hubspot fail on my end.

File: integrations/lib/multiwoven/integrations/source/aws_athena/client.rb [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+module Multiwoven::Integrations::Source
+  module AWSAthena
```
Comment (karthik-sivadas) [inline integrations/lib/multiwoven/integrations/source/aws_athena/client.rb:6]: While naming the class or module we follow the camel case naming convention based on the folder naming. If the folder name is `aws_athena` then the module name will be `AwsAthena`. Please update this accordingly

File: integrations/lib/multiwoven/integrations/source/aws_athena/client.rb [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+        ConnectionStatus.new(status: ConnectionStatusType["succeeded"]).to_multiwoven_message
+      rescue StandardError => e
+        ConnectionStatus.new(status: ConnectionStatusType["failed"], message: e.message).to_multiwoven_message
+      end
+
+      def discover(connection_config)
+        connection_config = connection_config.with_indifferent_access
+        query = "SELECT table_name, column_name, data_type, is_nullable FROM information_schema.columns WHERE table_schema = '#{connection_config[:schema]}' ORDER BY table_name, ordinal_position;"
+        db = create_connection(connection_config)
+        response = db.start_query_execution(
```
Comment (karthik-sivadas) [inline integrations/lib/multiwoven/integrations/source/aws_athena/client.rb:22]: We can make this DRY by extracting this into a function as this is used in read too. This will also increase the readability of the code

File: integrations/lib/multiwoven/integrations/source/aws_athena/client.rb [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+        end
+      end
+
+      def wait_for_query_completion(db, query_execution_id)
+        loop do
+          response = db.get_query_execution(query_execution_id: query_execution_id)
+          status = response.query_execution.status.state
+          break if %w[SUCCEEDED FAILED CANCELLED].include?(status)
+
+          sleep 1
```
Comment (karthik-sivadas) [inline integrations/lib/multiwoven/integrations/source/aws_athena/client.rb:84]: We refrain from using sleep.  Why do we need sleep and also do we need a loop?

Comment (TivonB-AI2) [reply integrations/lib/multiwoven/integrations/source/aws_athena/client.rb:84]: As for the sleep sometimes the get_query_execution keeps running instead of awaiting for a response. Will look into how to fix the problem without sleep

File: integrations/lib/multiwoven/integrations/source/aws_athena/client.rb (hunk 9 above)
Comment (afthabvp) [reply integrations/lib/multiwoven/integrations/source/aws_athena/client.rb:3]: remove this require "aws-sdk-athena" from client.rb
its already added in integrations.rb

File: integrations/lib/multiwoven/integrations/source/aws_athena/client.rb [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+        transform_results(db.get_query_results(query_execution_id: query_execution_id))
+      end
+
+      def create_streams(records)
+        group_by_table(records).map do |_, r|
+          Multiwoven::Integrations::Protocol::Stream.new(name: r[:tablename], action: StreamAction["fetch"], json_schema: convert_to_json_schema(r[:columns]))
+        end
+      end
+
+      def wait_for_query_completion(db, query_execution_id)
```
Comment (afthabvp) [inline integrations/lib/multiwoven/integrations/source/aws_athena/client.rb:76]: wait_for_query_completion is not used.please correct this

File: integrations/lib/multiwoven/integrations/rollout.rb
Comment (afthabvp) [inline integrations/lib/multiwoven/integrations/rollout.rb:1]: upgrade the rollout VERSION

**Review by karthik-sivadas — APPROVED** [review]

**Review by afthabvp — APPROVED** [review]

**Review by subintp — APPROVED** [review]
